User manager in meta

boshengchen opened the pull request:
Instead of PR #257 .
1. Addressed comments.
2. storage the data using dataman .

darionyaphet commented on src/meta/processors/AuthenticationProcessor.cpp:
```diff
@@ -0,0 +1,310 @@
+/* Copyright (c) 2018 vesoft inc. All rights reserved.
```
please update to `2019`

boshengchen replied:
Sorry, my bad, let me check all new files.

boshengchen replied:
I have checked the license head, and all the files are 2018 currently. Do we need to be changed them to 2019?

boshengchen replied:
Changed to 2019 with all new files as below :
src/meta/processors/AuthenticationProcessor.cpp
src/meta/processors/AuthenticationProcessor.h
src/meta/test/AuthProcessorTest.cpp

darionyaphet commented on src/interface/meta.thrift:
```diff
@@ -77,6 +89,22 @@ struct EdgeItem {
+    2: string first_name,
+    3: string last_name,
+    4: string email,
+    5: string phone,
```
why supported `phone` ?

boshengchen replied:
> why supported `phone` ?

It's just a user property. It is optional. More such options may be added in the future.

darionyaphet replied:
marked as `optional` will be better ?

boshengchen replied:
I think It is safe if set the optional to default (i.e. remove the optional keyword). What are the advantages of setting this parameter to `optional`?

darionyaphet replied:
I don't think these attributes are necessary.

dangleptr replied:
optional and default are not the same. 
default means the field must be written out, but we could read it or not as needs.  If the filed is not set,  we will read a default value for it. 
optional means if the field is not set,  we will read a nullptr.

In this case,  default and optional both are ok.  All of them are backward compatible. 
But for email and phone,  i think we have no needs for them.  But there are some other options we could support.   See https://dev.mysql.com/doc/refman/8.0/en/create-user.html

dangleptr commented on src/interface/meta.thrift:
```diff
@@ -291,6 +319,63 @@ struct HBReq {
+struct CreateUserReq {
+    1: UserItem user,
+    2: string encoded_pwd,
+    3: bool missing_ok,
```
What's this?

boshengchen replied:
> What's this?

missing_ok？ If the syntax contains `IF NOT EXISTS` ,  missing_ok should be true, else should be false.

dangleptr commented on src/meta/MetaServiceUtils.cpp:
```diff
@@ -301,5 +306,129 @@ cpp2::ErrorCode MetaServiceUtils::alterColumnDefs(std::vector<nebula::cpp2::Colu
+
+std::string MetaServiceUtils::userVal(const std::string& password,
+                                      const cpp2::UserItem& userItem,
+                                      std::shared_ptr<const SchemaProviderIf> schema) {
```
Don't use dataman in Meta module.  It is too heavy.

boshengchen commented:
Thanks for your response @darionyaphet @dangleptr , I have two questions to discuss as below :
1, About user options,  In fact, my original design and implementation was based on mysql ，There are options for default roles at that time. It was changed to what it is now after a couple of iterations, Thanks for your suggestions during the design phase. About option `email` and `phone`, The implementation of the syntax has been merged into the mainline，seems that we need to get a confirmed design again . and then re-implement these confirmed design. So,  Please feel free to let me know if have any good ideas . Let's using issue #190 for recording. 
2, About dataman , what are some good ways to avoid using dataman in module meta?  Please refer to #190  for requirements. Thanks.

boshengchen commented:
Address some comments as below:
1, Changed user data transport protocol from dataman to thrift. 
2, Removed user property email and phone. 
3, Added new user property lock status, this option is referenced from mysql.

dangleptr commented on src/interface/meta.thrift:
```diff
@@ -89,6 +101,21 @@ struct EdgeItem {
+
+struct UserItem {
+    1: string account,
+    2: string first_name,
```
Do we need first_name, last_name?   
Why not put password into this structure?

boshengchen replied:
Thanks for your review @dangleptr , here are the answers to each question:
1: About first_name and last_name, Discussed with  @sherman-the-tank in issue #190 , Need to add more properties to a user, such as email address, first name, last name, etc.
2:About password is not contained in the structure UserItem，the reason is functional difference is between password and other user attributes. password is only used when connecting or changing password, so password is not normal user attributes. Also, fast read and write do not require serialization or deserialization of structures.

dangleptr replied:
> 1: About first_name and last_name, Discussed with @sherman-the-tank in issue #190 , Need to add more properties to a user, such as email address, first name, last name, etc.

We'd better just saved the information our system used.  Just supposed that, when you create a account,  the system ask you to fill your personal information,  is it reasonable?

> 2:About password is not contained in the structure UserItem，the reason is functional difference is between password and other user attributes. password is only used when connecting or changing password, so password is not normal user attributes. Also, fast read and write do not require serialization or deserialization of structures.

Make sense.

dangleptr commented on src/interface/meta.thrift:
```diff
@@ -89,6 +101,21 @@ struct EdgeItem {
+}
+
+struct RoleItem {
+    1: string account,
```
Why not use userID?

dangleptr commented on src/interface/meta.thrift:
```diff
@@ -300,6 +327,72 @@ struct HBReq {
+}
+
+struct DropUserReq {
+    1: string account,
```
If you don't use "userId" in all requests, why we need it?

dangleptr commented on src/interface/meta.thrift:
```diff
@@ -89,6 +101,21 @@ struct EdgeItem {
+
+struct RoleItem {
+    1: string account,
+    2: string space,
```
why not spaceId?

dangleptr commented on src/interface/meta.thrift:
```diff
@@ -43,10 +46,19 @@ enum AlterSchemaOp {
 
 
+enum RoleType {
+    GOD    = 0x01,
```
Add some comments, what's the meaning for different roles?

boshengchen replied:
Good suggest!

dangleptr commented on src/interface/meta.thrift:
```diff
@@ -300,6 +327,72 @@ struct HBReq {
+}
+
+struct ListRolesReq {
+    1: string space,
```
ditto

boshengchen commented:
About SpaceID and UserID exists multiple places, I've actually thought about it，Because these two structures are used for user management often. So  can use these names directly . I think convert
 from name to id is superfluous in here. What do you think?
UserID and SpaceID are actually useful, they will be used for privilege management. I will use UserID and SpaceID to implement some caches in meta server or meta client.

dangleptr commented:
> About SpaceID and UserID exists multiple places, I've actually thought about it，Because these two structures are used for user management often. So can use these names directly . I think convert
> from name to id is superfluous in here. What do you think?
> UserID and SpaceID are actually useful, they will be used for privilege management. I will use UserID and SpaceID to implement some caches in meta server or meta client.

It make sense for userId,  but not for spaceID.  We'd better use spaceId to join different entries instead of spaceName.

dangleptr commented:
By the way,  please refer #483 .  Put the user related processors under a separate dir.

boshengchen commented:
> It make sense for userId, but not for spaceID. We'd better use spaceId to join different entries instead of spaceName.
That makes sense, let me improve the space name and RoleItem related.

boshengchen commented:
1,Refactor user processors code structure. 
2,Remove user properties first name and last name. 
3, Add new user properties of resource limit. 
4, Changed related thrift structure from `'name'` to `'id'`, 
5,Add some comments for role.

boshengchen commented:
Ready to review. Please review again. Thanks.

dangleptr commented on src/graph/PermissionManager.h:
```diff
@@ -0,0 +1,47 @@
+#ifndef GRAPH_PERMISSIONMANAGER_H
+#define GRAPH_PERMISSIONMANAGER_H
+
+// Operation and permission define:
```
Well done

dangleptr commented on src/meta/processors/BaseProcessor.inl:
```diff
@@ -162,6 +162,18 @@ Status BaseProcessor<RESP>::spaceExist(GraphSpaceID spaceId) {
 }
 
+template<typename RESP>
+Status BaseProcessor<RESP>::userExist(UserID spaceId) {
```
Could we unify the userExist with getUserId ?

boshengchen replied:
`userExist` is judged by UserID , But  `getUserId` is  judged by userName. And they query for different tables userIndex and userData. They are same with spaceExist and getSpaceId. I think it's best not to unify them. WDYT?

dangleptr replied:
Fair enough.

dangleptr commented on src/meta/processors/partsMan/DropSpaceProcessor.cpp:
```diff
@@ -41,6 +41,16 @@ void DropSpaceProcessor::process(const cpp2::DropSpaceReq& req) {
     deleteKeys.emplace_back(MetaServiceUtils::indexSpaceKey(req.get_space_name()));
     deleteKeys.emplace_back(MetaServiceUtils::spaceKey(spaceId));
 
+    // delete related role data.
```
Now we just delete the parts information when deleting space.  We could delete the useless user and schema when compaction.

boshengchen replied:
Yeah, that makes sense.  Let me modify it later.

dangleptr commented on src/meta/processors/usersMan/AuthenticationProcessor.cpp:
```diff
@@ -0,0 +1,271 @@
+}
+
+
+void AlterUserProcessor::process(const cpp2::AlterUserReq& req) {
```
It seems the alter is to change the whole value. 
I want to know what's the language for alter user?

boshengchen replied:
No, Only change properties that require alter.

**Alter user sentences:**
```
ALTER USER account WITH with_user_opt_list

with_user_opt_item
    : KW_ACCOUNT KW_LOCK
    | KW_ACCOUNT KW_UNLOCK
    | KW_MAX_QUERIES_PER_HOUR INTEGER
    | KW_MAX_UPDATES_PER_HOUR INTEGER
    | KW_MAX_CONNECTIONS_PER_HOUR INTEGER
    | KW_MAX_USER_CONNECTIONS INTEGER
    ;

with_user_opt_list
    : with_user_opt_item
    | with_user_opt_list COMMA with_user_opt_item
    ;
```

dangleptr commented on src/meta/processors/usersMan/AuthenticationProcessor.cpp:
```diff
@@ -0,0 +1,271 @@
+
+
+void ChangePasswordProcessor::process(const cpp2::ChangePasswordReq& req) {
+    folly::SharedMutex::WriteHolder wHolder(LockUtils::userLock());
```
Change pwd is a special sentence?

boshengchen replied:
Yes, Change password is a separate sentence. No need to verify the password if it is a `GOD` role. else the password must be validated. And `GOD` can change anyone's password. Other roles can only change their own password.

```
change_password_sentence
    : CHANGE PASSWORD account TO new_password
    | CHANGE PASSWORD account FROM old_password TO new_password
    ;
```

dangleptr commented on src/meta/processors/usersMan/AuthenticationProcessor.cpp:
```diff
@@ -0,0 +1,271 @@
+        return;
+    }
+
+    if (!checkPassword(userRet.value(), req.get_encoded_pwd())) {
```
I want to know when to check the password?

boshengchen replied:
This is a reserved interface，Used for client or terminal connections.

boshengchen commented:
To improve the drop space , don't delete useless role data.

dangleptr commented:
Jenkins go

boshengchen commented:
1, Rebased 
2, Resolved compilation errors with depends on  `raftex_obj raftex_thrift_obj wal_obj time_obj`.
Please review again . Thanks.

boshengchen commented:
Rebased and resolved some conflicts.

laura-ding commented on src/meta/processors/usersMan/AuthenticationProcessor.cpp:
```diff
@@ -0,0 +1,271 @@
+
+    // Collect related roles by user.
+    std::unique_ptr<kvstore::KVIterator> iter;
+    std::string prefix = "__roles__";
```
I think the prefix should be used kRolesTable although it define in cpp file

boshengchen replied:
Good ! We are think alike, I don't feel comfortable using it here either. Let me try.

laura-ding commented on src/meta/test/CMakeLists.txt:
```diff
@@ -171,3 +171,33 @@ nebula_link_libraries(
+        $<TARGET_OBJECTS:raftex_obj>
+        $<TARGET_OBJECTS:raftex_thrift_obj>
+        $<TARGET_OBJECTS:wal_obj>
+	$<TARGET_OBJECTS:thrift_obj>
```
Alignment ：185L and 189L

boshengchen replied:
Sharp and bright eyes

laura-ding commented on src/meta/test/TestUtils.h:
```diff
@@ -184,6 +185,37 @@ class TestUtils {
         return sc;
     }
+
+    static StatusOr<UserID> createUser(kvstore::KVStore* kv,
```
Strange alignment

boshengchen commented:
Address laura-ding's comment

laura-ding commented on src/meta/test/CMakeLists.txt:
```diff
@@ -182,11 +182,11 @@ add_executable(
         $<TARGET_OBJECTS:raftex_thrift_obj>
         $<TARGET_OBJECTS:wal_obj>
-	$<TARGET_OBJECTS:thrift_obj>
+	    $<TARGET_OBJECTS:thrift_obj>
```
I see it's still not aligned

boshengchen replied:
My bad, resolved.  Thanks for your careful review @laura-ding

dangleptr commented:
Jenkins go

laura-ding commented on src/meta/test/CMakeLists.txt:
```diff
@@ -171,3 +171,33 @@ nebula_link_libraries(
     gtest
 )
 nebula_add_test(meta_http_test)
+add_executable(
```
![image](https://user-images.githubusercontent.com/48548375/59732446-d3e34e00-927c-11e9-9f3d-1468d0a03175.png)
There are no aligned with 4 Spaces

boshengchen replied:
A stupid mistake，corrected.

boshengchen commented:
Jenkins go

dangleptr commented:
Jenkins go
